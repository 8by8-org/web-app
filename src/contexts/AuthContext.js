import React, { useContext, useEffect, useState } from "react";
import { getUserDatabaseByUID } from "../functions/UserData";
import { auth } from "../firebase";

export function useAuth() {
  return useContext(AuthContext);
}

export const AuthContext = React.createContext();

export function AuthProvider({ children }) {
  const [currentUser, setCurrentUser] = useState(null);
  //keep track of user data in the context to limit calls to firestore
  const [currentUserData, setCurrentUserData] = useState(null);
  const [authLoading, setAuthLoading] = useState(true);
  //error handling
  const [error, setError] = useState(null);
  /*keep track of voter reg form data here so that the flow can be spread out over several
  discrete pages*/
  const [voterRegistrationData, setVoterRegistrationData] = useState({
    send_confirmation_reminder_emails: false,
    dob: "",
    idNumber: "",
    email: "",
    first_registration: false,
    zip: "",
    citizen: false,
    has_state_license: false,
    eighteenPlus: false,
    name_title: "",
    name_first: "",
    name_middle: "",
    name_last: "",
    street: "",
    streetLine2: "",
    unit: "",
    city: "",
    state: "AL",
    has_mailing_address: false,
    mailing_street: "",
    mailing_streetLine2: "",
    mailing_unit: "",
    mailing_city: "",
    mailing_state: "AL",
    mailing_zip: "",
    phone: "Home",
    change_of_name: false,
    prev_name_title: "",
    prev_name_first: "",
    prev_name_middle: "",
    prev_name_last: "",
    change_of_address: false,
    prev_street: "",
    prev_streetLine2: "",
    prev_unit: "",
    prev_city: "",
    prev_state: "AL",
    prev_zip: "",
    opt_in_email: false,
    opt_in_sms: false,
    opt_in_volunteer: false,
    party: "",
    race: "",
  });

  useEffect(() => {
    const unsubscribe = auth.getAuth().onAuthStateChanged((user) => {
      setError(null);
      setCurrentUser(user);
      //refresh voter registration data
      setVoterRegistrationData({
        dob: "",
        idNumber: "",
        email: user ? user.email : "",
        first_registration: false,
        zip: "",
        citizen: false,
        has_state_license: false,
        eighteenPlus: false,
        title: "",
        name_first: "",
        name_middle: "",
        name_last: "",
        suffix: "",
        street: "",
        streetLine2: "",
        unit: "",
        city: "",
        state: "AL",
        diff_mailing_address: false,
        mail_street: "",
        mail_streetLine2: "",
        mail_unit: "",
        mail_city: "",
        mail_state: "AL",
        mail_zip: "",
        change_of_name: false,
        prev_title: "",
        prev_name_first: "",
        prev_name_middle: "",
        prev_name_last: "",
        prev_suffix: "",
        change_of_address: false,
        prev_street: "",
        prev_streetLine2: "",
        prev_unit: "",
        prev_city: "",
        prev_state: "AL",
        prev_zip: "",
        party: "",
        race: "",
      });
      //if a user has been found, get the current user's information from firestore
      if (user) {
        getUserDatabaseByUID(user.uid)
          .then((userData) => {
            setCurrentUserData(userData);
            setAuthLoading(false);
          })
          .catch((e) => {
            //set an error state to the error message, and set authLoading to false
            setError(e.message);
            setAuthLoading(false);
          });
      } else {
        //if no user was found, set currentUserData to null and authLoading to false
        setCurrentUserData(null);
        setAuthLoading(false);
      }
    });
    return unsubscribe;
  }, []);

  const value = {
    currentUser,
    currentUserData,
    setCurrentUserData,
    voterRegistrationData,
    setVoterRegistrationData,
  };
  return (
    <AuthContext.Provider value={value}>
      {!authLoading && children}
    </AuthContext.Provider>
  );
}
