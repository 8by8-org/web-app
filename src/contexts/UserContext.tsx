import {createContext, PropsWithChildren, useState } from "react";
import { User } from "@/models/User";
import { Builder } from "builder-pattern";
import { UserType } from "@/models/UserType";

type UserContextType = {
  activeUser:ActiveUser;
}

type ActiveUser = User | null;

//setting the default context to {activeUser:null} allows for testing unwrapped components
export const UserContext = createContext<UserContextType>({activeUser: null});

export function UserContextProvider({children}:PropsWithChildren) {
  const user = Builder(User).name("Joe").type(UserType.CHALLENGER).build();
  
  const [activeUser,setActiveUser] = useState<ActiveUser>(user);

  return <UserContext.Provider value={{activeUser}}>{children}</UserContext.Provider>
}

