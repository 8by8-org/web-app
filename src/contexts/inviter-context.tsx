import { Inviter } from "@/models/Inviter";
import { createContext, PropsWithChildren, useState } from "react";

export type InviterContextType = {
  inviterInfo:InviterInfo;
}

type InviterInfo = Inviter | null;

//setting the default context to {activeUser:null} allows for testing unwrapped components
export const InviterContext = createContext<InviterContextType>({inviterInfo: null});

export function InviterContextProvider({children}:PropsWithChildren) {
  const [inviterInfo,setInviterInfo] = useState<InviterInfo>(null);
  return <InviterContext.Provider value={{inviterInfo}}>{children}</InviterContext.Provider>
}

