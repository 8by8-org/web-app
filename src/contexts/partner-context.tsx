import { createContext, PropsWithChildren, useState } from "react";

export type PartnerContextType = {
  partnersExist: boolean;
};

//setting the default context to {partnersExist:false} allows for testing unwrapped components
export const PartnerContext = createContext<PartnerContextType>({
  partnersExist: false,
});

export function PartnerContextProvider({ children }: PropsWithChildren) {
  const [partnersExist, setPartnersExist] = useState(false);
  return (
    <PartnerContext.Provider value={{ partnersExist }}>
      {children}
    </PartnerContext.Provider>
  );
}
