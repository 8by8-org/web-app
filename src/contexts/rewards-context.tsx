import { Rewards } from "@/models/Rewards";
import { createContext, PropsWithChildren, useState } from "react";

export type RewardsContextType = {
    rewardsInfo:RewardsInfo;
}

type RewardsInfo = Rewards | null;

//setting the default context to {rewardsInfo:null} allows for testing unwrapped components
export const RewardsContext = createContext<RewardsContextType>({rewardsInfo: null});

export function RewardsContextProvider({children}:PropsWithChildren) {
  const [rewardsInfo,setRewardsInfo] = useState<RewardsInfo>(null);
  return <RewardsContext.Provider value={{rewardsInfo}}>{children}</RewardsContext.Provider>
}

