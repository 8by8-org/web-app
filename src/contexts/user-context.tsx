import {createContext, PropsWithChildren, useState } from "react";
import { User } from "@/models/User";
import { UserType } from "@/models/UserType";

export type UserContextType = {
  activeUser:ActiveUser;
}

type ActiveUser = User | null;

//setting the default context to {activeUser:null} allows for testing unwrapped components
export const UserContext = createContext<UserContextType>({activeUser: null});

export function UserContextProvider({children}:PropsWithChildren) {
  const [activeUser,setActiveUser] = useState<ActiveUser>(null);
  return <UserContext.Provider value={{activeUser}}>{children}</UserContext.Provider>
}

