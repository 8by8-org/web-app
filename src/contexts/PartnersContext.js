import React, {useState, useEffect, createContext, useContext} from "react";
import { getAllPartnerData } from "../functions/partnerData";
import { useAuth } from "./AuthContext";

export function usePartners() {
    return useContext(PartnersContext);
}

export const PartnersContext = createContext();

export const PartnersContextProvider = ({children}) => {
    const {currentUser} = useAuth();
    const [allPartners, setAllPartners] = useState({});
    const [partnersExist, setPartnersExist] = useState(false);

    function setPartnersAndPartnersExists(data) {
        setAllPartners(data);
        if(data && Object.keys(data).length > 0) setPartnersExist(true);
        else setPartnersExist(false);
    }

    //this should also fire on auth state change
    useEffect(() => {
        getAllPartnerData(setPartnersAndPartnersExists);
    }, []);

    useEffect(() => {
        getAllPartnerData(setPartnersAndPartnersExists);
    }, [currentUser]);

    return (
        <PartnersContext.Provider value={{allPartners, partnersExist}}>{children}</PartnersContext.Provider>
    )
}