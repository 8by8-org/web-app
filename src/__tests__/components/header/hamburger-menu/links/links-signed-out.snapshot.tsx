import { render, cleanup } from '@testing-library/react'
import {Builder} from 'builder-pattern';
import { UserContextType, UserContext } from '@/contexts/user-context';
import Links from '@/components/header/hamburger-menu/links/links';

describe('Links--Signed Out', () => {
  afterEach(cleanup);

  it('renders signed out links unchanged', () => {
    const userContextValue = Builder<UserContextType>().activeUser(null).build();
    const {container} = render(
      <UserContext.Provider value={userContextValue}>
        <Links />
      </UserContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });
});