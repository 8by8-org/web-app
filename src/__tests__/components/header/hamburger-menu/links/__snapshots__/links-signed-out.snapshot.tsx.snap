// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Links--Signed Out renders signed out links unchanged 1`] = `
<div>
  <li>
    <a
      class="take_the_challenge_btn"
      href="/challengerwelcome"
    >
      Take The Challenge
    </a>
  </li>
  <li>
    <a
      class="link_lg_top"
      href="/actions"
    >
      Take Action
    </a>
  </li>
  <li>
    <a
      class="link_lg"
      href="/why8by8"
    >
      Why 8by8
    </a>
  </li>
  <li>
    <a
      class="link_lg"
      href="/rewards"
    >
      Rewards
    </a>
  </li>
  <li>
    <a
      class="link_lg"
      href="/faq"
    >
      FAQS
    </a>
  </li>
  <li>
    <a
      class="link_sm_top"
      href="/privacy"
    >
      Privacy Policy
    </a>
  </li>
  <li>
    <a
      class="link_sm"
      href="/settings"
    >
      Settings
    </a>
  </li>
  <li>
    <a
      class="link_sm"
      href="/signup"
    >
      Sign up
    </a>
  </li>
</div>
`;
